lib/Runtime/Language/TaggedInt.cpp

```diff
@@ -428,21 +428,65 @@ namespace Js
         return JavascriptNumber::ToVar(uValue >> (nShift & 0x1F), scriptContext);
     }
 
-    void TaggedInt::ToBuffer(Var aValue, __out_ecount_z(bufSize) char16 * buffer, uint bufSize)
+    int TaggedInt::ToBuffer(Var aValue, __out_ecount_z(bufSize) char16 * buffer, uint bufSize)
     {
         return ToBuffer(ToInt32(aValue), buffer, bufSize);
     }
 
-    void TaggedInt::ToBuffer(int value, __out_ecount_z(bufSize) char16 * buffer, uint bufSize)
+    // fills the buffer from the end and returns the start index
+    static int UnsignedToString(unsigned long value, char16 *buffer, int bufferSize)
+    {
+        buffer[bufferSize - 1] = char16(0);
+        int pos = bufferSize - 2;
+        while(value > 9 && pos > 1)
+        {
+            const int val100 = value % 100;
+            value /= 100;
+
+            if (val100 < 10)
+            {
+                buffer[pos--] = _u('0') + static_cast<char>(val100);
+                buffer[pos--] = _u('0');
+                continue;
+            }
+
+            buffer[pos--] = _u('0') + static_cast<char>(val100 % 10);
+            buffer[pos--] = _u('0') + static_cast<char>(val100 / 10);
+        }
+
+        if (pos == 0) return 0; // well, the buffer wasn't enough. just return whatever we had
+
+        if (value && value < 10)
+        {
+            buffer[pos--] = _u('0') + static_cast<char>(value);
+        }
+        else if (pos == bufferSize - 2) // if it was 0
+        {
+            buffer[pos--] = _u('0');
+        }
+
+        return pos + 1;
+    }
+
+    static int SignedToString(long value, char16 *buffer, int bufferSize)
+    {
+        bool neg = value < 0;
+        unsigned long val = (unsigned long) (neg ? -1 * (long)value : value);
```

Comment:
Overflow and such are undefined behavior for signed ints. As -LONG_MIN is outside the range of a long, it's UB, but in a practical sense it _should_ work as expected on all architectures that we support at the moment. The biggest danger is an aggressive compiler assuming that the resultant `val` has a 0 high-bit, and omitting some range check, but since we don't have any such checks here, it should be non-issue.